receiver/carbonreceiver/transport/tcp_server.go

```diff
@@ -0,0 +1,213 @@
+// Copyright 2019, OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package transport
+
+import (
+	"bufio"
+	"context"
+	"fmt"
+	"io"
+	"net"
+	"strings"
+	"sync"
+	"time"
+
+	metricspb "github.com/census-instrumentation/opencensus-proto/gen-go/metrics/v1"
+	"github.com/open-telemetry/opentelemetry-collector/consumer"
+	"github.com/open-telemetry/opentelemetry-collector/consumer/consumerdata"
+	"go.opencensus.io/trace"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/receiver/carbonreceiver/protocol"
+)
+
+const (
+	// TCPIdleTimeoutDefault is the default timeout for idle TCP connections.
+	TCPIdleTimeoutDefault = 30 * time.Second
+)
+
+type tcpServer struct {
+	ln          net.Listener
+	wg          sync.WaitGroup
+	idleTimeout time.Duration
+	reporter    Reporter
+}
+
+var _ (Server) = (*tcpServer)(nil)
+
+// NewTCPServer creates a transport.Server using TCP as its transport.
+func NewTCPServer(
+	addr string,
+	idleTimeout time.Duration,
+) (Server, error) {
+	if idleTimeout < 0 {
+		return nil, fmt.Errorf("invalid idle timeout: %v", idleTimeout)
+	}
+
+	if idleTimeout == 0 {
+		idleTimeout = TCPIdleTimeoutDefault
+	}
+
+	ln, err := net.Listen("tcp", addr)
+	if err != nil {
+		return nil, err
+	}
+
+	t := tcpServer{
+		ln:          ln,
+		idleTimeout: idleTimeout,
+	}
+	return &t, nil
+}
+
+func (t *tcpServer) ListenAndServe(
+	parser protocol.Parser,
+	nextConsumer consumer.MetricsConsumer,
+	reporter Reporter,
+) error {
+	if parser == nil || nextConsumer == nil || reporter == nil {
+		return errNilListenAndServeParameters
+	}
+
+	acceptedConnMap := make(map[net.Conn]struct{})
+	connMapMtx := &sync.Mutex{}
+
+	t.reporter = reporter
+	var err error
+	for {
+		conn, acceptErr := t.ln.Accept()
+		if acceptErr == nil {
+			connMapMtx.Lock()
+			acceptedConnMap[conn] = struct{}{}
+			connMapMtx.Unlock()
+			t.wg.Add(1)
+			go func(c net.Conn) {
+				t.handleConnection(parser, nextConsumer, c)
+				connMapMtx.Lock()
+				delete(acceptedConnMap, c)
+				connMapMtx.Unlock()
+				t.wg.Done()
+			}(conn)
+			continue
+		}
+
+		if netErr, ok := acceptErr.(net.Error); ok {
+			t.reporter.OnDebugf(
+				"TCP Transport (%s) - Accept (temporary=%v) net.Error: %v",
+				t.ln.Addr().String(),
+				netErr.Temporary(),
+				netErr)
+			if netErr.Temporary() {
+				continue
+			}
+		}
+
+		err = acceptErr
+		break
```

Comment:
Is it only possible to exit this loop with an error? Is there a way to do errorless shutdown of this receiver without printing of error in the log?

Reply:
It will print error only if debug is enabled, in which case it will print an error indicating use of closed connection.